### PR TITLE
QueryServer: Add support for new query-server

### PR DESCRIPTION
**Pull request by alexet**
Add support for the new query server.

For local testing purposes it hardcodes cli support to true. We need to change to the correct version before merging.

### CODE REVIEW

File: extensions/ql-vscode/src/databases-ui.ts [hunk 1]
```diff
@@ -390,9 +388,9 @@ export class DatabaseUI extends DisposableObject {
...
     progress: ProgressCallback,
     token: CancellationToken
-  ): Promise<DatabaseItem | undefined> => {
+  ): Promise<void> => {
     try {
-      return await this.chooseAndSetDatabase(true, progress, token);
+      await this.chooseAndSetDatabase(true, progress, token);
     } catch (e) {
       void showAndLogErrorMessage(getErrorMessage(e));
       return undefined;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/databases-ui.ts:396]: This line is not necessary.

File: extensions/ql-vscode/src/extension.ts [hunk 2]
```diff
@@ -70,12 +70,12 @@ import { asError, assertNever, getErrorMessage } from './pure/helpers-pure';
 import { spawnIdeServer } from './ide-server';
 import { InterfaceManager } from './interface';
 import { WebviewReveal } from './interface-utils';
-import { ideServerLogger, logger, queryServerLogger } from './logging';
+import { ideServerLogger, logger, ProgressReporter, queryServerLogger } from './logging';
 import { QueryHistoryManager } from './query-history';
 import { CompletedLocalQueryInfo, LocalQueryInfo } from './query-results';
-import * as qsClient from './queryserver-client';
+import * as qsClient from './legacy-query-server/queryserver-client';
+import * as qsClient2 from './query-server/queryserver-client';
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/extension.ts:77]: Maybe this is more understandable?

```suggestion
import * as legacyQsClient from './legacy-query-server/queryserver-client';
import * as newQsClient from './query-server/queryserver-client';
```

File: extensions/ql-vscode/src/extension.ts [hunk 3]
```diff
@@ -782,12 +770,13 @@ async function activateWithInstalledDistribution(
           });
         }
 
-        if (queryUris.length > 1) {
+        if (queryUris.length > 1 && !await cliServer.cliConstraints.supportsNonDestructiveUpgrades()) {
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/extension.ts:773]: Interesting...we never had this before. Probably a bug that we always tried to upgrade the database. Most likely in these cases, this was a no-op that just took some time.

File: extensions/ql-vscode/src/legacy-query-server/upgrades.ts [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
-import { getOnDiskWorkspaceFolders, showAndLogErrorMessage, tmpDir } from './helpers';
-import { ProgressCallback, UserCancellationException } from './commandRunner';
-import { logger } from './logging';
-import * as messages from './pure/messages';
-import * as qsClient from './queryserver-client';
+import { getOnDiskWorkspaceFolders, showAndLogErrorMessage, tmpDir } from '../helpers';
+import { ProgressCallback, UserCancellationException } from '../commandRunner';
+import { logger } from '../logging';
+import * as messages from '../pure/legacy-messages';
+import * as qsClient from '../legacy-query-server/queryserver-client';
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/legacy-query-server/upgrades.ts:6]: Unnecessary change?

File: extensions/ql-vscode/src/pure/new-messages.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+
+
+export interface UpgradeParams {
+  db: string,
+  additionalPacks: string[],
+}
+
+export type UpgradeResult = Record<string, unknown>;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/pure/new-messages.ts:150]: Surprised we still need upgrades here.

File: extensions/ql-vscode/src/query-server/query-runner.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  }
+  async registerDatabase(progress: ProgressCallback, token: CancellationToken, dbItem: DatabaseItem): Promise<void> {
+    if (dbItem.contents && (await this.qs.cliServer.cliConstraints.supportsDatabaseRegistration())) {
+      const databases: string[] = [dbItem.databaseUri.fsPath];
+      await this.qs.sendRequest(registerDatabases, { databases }, token, progress);
+    }
+  }
+
+  async upgradeDatabaseExplicit(dbItem: DatabaseItem, progress: ProgressCallback, token: CancellationToken): Promise<void> {
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/query-server/query-runner.ts:59]: Ah... I see. Is this only for very old databases? Maybe include a note saying something like:

> Databases created by CLI versions later than 2.x.y do not need to be explicitly upgraded.

File: extensions/ql-vscode/src/run-queries-shared.ts [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+   */
+  async ensureDilPath(cliServer: CodeQLCliServer): Promise<string> {
+    if (await this.hasDil()) {
+      return this.dilPath;
+    }
+    const compiledQuery = path.join(this.querySaveDir, 'compiledQuery.qlo');
+    if (!(await fs.pathExists(compiledQuery))) {
+      if (await cliServer.cliConstraints.supportsNewQueryServer()) {
+        // This could be from the new query server
+        // in whcih case we expect the qlo to be missing so we should ignore it
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-queries-shared.ts:199]:
```suggestion
        // in which case we expect the qlo to be missing so we should ignore it
```

File: extensions/ql-vscode/src/run-queries-shared.ts [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+        throw new Error(
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-queries-shared.ts:200]: When would this happen? Is there anything the user can do to fix this situation? The error message as it is is not very helpful.

**Review by aeisenberg — COMMENTED** [review]
Looks generally reasonable. I haven't tried this out myself.

**Comment by aeisenberg** [comment]
I think you want to base this PR on top of https://github.com/github/vscode-codeql/pull/1498, which might be difficult since that PR is from your local fork.

File: extensions/ql-vscode/src/query-server/query-runner.ts (hunk 6 above)
Comment (alexet) [reply extensions/ql-vscode/src/query-server/query-runner.ts:59]: This is because there are a couple of minor reasons why we may want an explicit upgrade. Old databases don't actually work but there are still some small reasons why you may want a destructive upgrade (usually performance testing).

File: extensions/ql-vscode/src/run-queries-shared.ts (hunk 8 above)
Comment (alexet) [reply extensions/ql-vscode/src/run-queries-shared.ts:200]: Originally I though that this should not happen unless the user deleted the dil or a bug happened in the query server. But I have also realised that a compilation error would also cause this. It is a bit tricky to work out which case but hopefully it is understandable that if your query doesn't compile then you won't be able to see the DIL.

**Comment by aeisenberg** [comment]
Superseded by https://github.com/github/vscode-codeql/pull/1508

**Comment by aeisenberg** [comment]
Closing in favour of #1508.
